providers/implementations/encode_decode/ml_kem_codecs.c

```diff
@@ -0,0 +1,559 @@
+                continue;
+            ret[i].vp8_pref = ++count;
+            break;
+        }
+        fmt = end;
+    } while (count < NUM_PKCS8_FORMATS);
+
+    if (count == 0) {
+        OPENSSL_free(ret);
+        ERR_raise_data(ERR_LIB_PROV, PROV_R_ML_KEM_NO_FORMAT,
+                       "no %s private key %s formats are enabled",
+                       algorithm_name, direction);
+        return NULL;
+    }
+    qsort(ret, NUM_PKCS8_FORMATS, sizeof(*ret), vp8_pref_cmp);
+    ret[count].vp8_entry = NULL;
+    return ret;
+}
+
+ML_KEM_KEY *
+ossl_ml_kem_d2i_PUBKEY(const uint8_t *pubenc, int publen, int evp_type,
+                       PROV_CTX *provctx, const char *propq)
+{
+    OSSL_LIB_CTX *libctx = PROV_LIBCTX_OF(provctx);
+    const ML_KEM_VINFO *v;
+    const ML_KEM_CINFO *c;
+    const ML_KEM_SPKI_INFO *vspki;
+    ML_KEM_KEY *ret;
+
+    if ((v = ossl_ml_kem_get_vinfo(evp_type)) == NULL
+        || (c = ml_kem_get_cinfo(evp_type)) == NULL)
+        return NULL;
+    vspki = c->spki_info;
+    if (publen != ML_KEM_SPKI_OVERHEAD + (ossl_ssize_t) v->pubkey_bytes
+        || memcmp(pubenc, vspki->asn1_prefix, ML_KEM_SPKI_OVERHEAD) != 0)
+        return NULL;
+    publen -= ML_KEM_SPKI_OVERHEAD;
+    pubenc += ML_KEM_SPKI_OVERHEAD;
+
+    if ((ret = ossl_ml_kem_key_new(libctx, propq, evp_type)) == NULL)
+        return NULL;
+
+    if (!ossl_ml_kem_parse_public_key(pubenc, (size_t) publen, ret)) {
+        ERR_raise_data(ERR_LIB_PROV, PROV_R_BAD_ENCODING,
+                       "errror parsing %s public key from input SPKI",
+                       v->algorithm_name);
+        ossl_ml_kem_key_free(ret);
+        return NULL;
+    }
+
+    return ret;
+}
+
+ML_KEM_KEY *
+ossl_ml_kem_d2i_PKCS8(const uint8_t *prvenc, int prvlen,
+                      int evp_type, PROV_CTX *provctx,
+                      const char *propq)
+{
+    OSSL_LIB_CTX *libctx = PROV_LIBCTX_OF(provctx);
+    const ML_KEM_VINFO *v;
+    const ML_KEM_CINFO *c;
+    ML_KEM_PKCS8_PREF *vp8_alloc = NULL, *vp8_slot;
+    const ML_KEM_PKCS8_INFO *vp8;
+    ML_KEM_KEY *key = NULL, *ret = NULL;
+    PKCS8_PRIV_KEY_INFO *p8inf = NULL;
+    const uint8_t *buf, *pos;
+    const X509_ALGOR *alg = NULL;
+    const char *formats;
+    int len, ptype;
+    uint32_t magic;
+    uint16_t seed_magic;
+
+    /* Which ML-KEM variant? */
+    if ((v = ossl_ml_kem_get_vinfo(evp_type)) == NULL
+        || (c = ml_kem_get_cinfo(evp_type)) == NULL)
+        return 0;
+
+    /* Extract the key OID and any parameters. */
+    if ((p8inf = d2i_PKCS8_PRIV_KEY_INFO(NULL, &prvenc, prvlen)) == NULL)
+        return 0;
+    /* Shortest prefix is 4 bytes: seq tag/len  + octet string tag/len */
+    if (!PKCS8_pkey_get0(NULL, &buf, &len, &alg, p8inf))
+        goto end;
+    /* Bail out early if this is some other key type. */
+    if (OBJ_obj2nid(alg->algorithm) != evp_type)
+        goto end;
+
+    /* Get the list of enabled decoders. Their order is not important here. */
```

Comment:
If the order is not important why do you even bother calling sorting them in vp8_order()?

Reply:
The order is important on output and useful to trim unsupported values on input.